src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20220608_194500.java

```diff
@@ -0,0 +1,119 @@
+        results.forEach(result -> {
+            if (result.getResultString() != null && result.getResultString().matches(".*of.*passed.*")) {
+                unprocessedResults.add(result);
+            }
+            else {
+                processedResults.add(result);
+            }
+        });
+
+        LOGGER.info("Found {} results to process.", unprocessedResults.size());
+
+        Lists.partition(results, 100).forEach(resultList -> {
+            LOGGER.info("Process (next) 100 results for the migration in one batch...");
+            processResults(resultList);
+        });
+
+        if (!unprocessedResults.isEmpty()) {
+            reportUnprocessedResults(unprocessedResults);
+        }
+    }
+
+    /*
+     * Takes a result string such as "21 of 42 passed, 42 issues" and splits it into its different components: The number of tests, the number of passed tests and the number of
+     * issues.
+     * @param results Batch of results that should get processed
+     */
+    private void processResults(List<Result> results) {
+        results.forEach(result -> {
+            String[] resultStringParts = result.getResultString().split(", ");
+            for (String resultStringPart : resultStringParts) {
+                // Matches e.g. "21 of 42 passed"
+                if (resultStringPart.matches(".*of.*passed.*")) {
+                    String[] testCaseParts = resultStringPart.split(" ");
+                    int passedTestCasesAmount = Integer.parseInt(testCaseParts[0]);
+                    int testCasesAmount = Integer.parseInt(testCaseParts[2]);
+
+                    result.setPassedTestCaseCount(passedTestCasesAmount);
+                    result.setTestCaseCount(testCasesAmount);
+                    result.setResultString(null);
+                }
+                // Matches e.g. "9 issues"
+                else if (resultStringPart.contains("issue")) {
+                    String[] issueParts = resultStringPart.split(" ");
+                    int codeIssueCount = Integer.parseInt(issueParts[0]);
+
+                    result.setCodeIssueCount(codeIssueCount);
+                    result.setResultString(null);
+                }
+            }
+        });
+
+        resultRepository.saveAll(results);
+    }
+
+    private void reportUnprocessedResults(List<Result> unprocessedResults) {
+        LOGGER.error("Found {} results of programming exercises that could not be processed! These results will no longer show a correct resultString!", unprocessedResults.size());
+        Optional<ZonedDateTime> latestUnprocessedResult = unprocessedResults.stream().map(Result::getCompletionDate).filter(Objects::nonNull).max(Comparator.naturalOrder());
+        latestUnprocessedResult.ifPresent(zonedDateTime -> LOGGER.error("The latest of these unprocessed results is from {}.", zonedDateTime));
+        LOGGER.error("The IDs of the unprocessed results are: {}", unprocessedResults.stream().map(Result::getId).toList());
```

Comment:
These numbers are only needed for the correct representation of the result string and not for the actual score calculation or something important. When the student pushes another submission it will get overwritten anyway and the points are also correctly shown in all cases, even if the migration fails for some results. Additionally the student can take a look at the feedback and see the test cases for themself.
So I don't see a reason why the application should not start if some results cannot get migrated

Reply:
From the student side, yes. Tutors+ can see all results though.

In general: What is a migration? A migration is the change of an entity from one state to another.
In this case, we have a structural migration of the database. Hence, we have one state (column resultString) and another state (multiple data columns).

It's up to you how you define the state but my definition would be that the resultString column should be all null to indicate a successful migration. Thus in the future, we can delete the column without a second thought.
Therefore, having non-null values in there is a violation of the state and should be fixed.
In the end, it could be a fault in the migration.

If you decide that non-null values in your state are okay and you only extract the values if they are available, then that is okay but we probably will have to do a second check in the future because unless we force this fallback, we never know for sure if everything will work on every instance that exists.

Of course, a good system admin will check the logs after an update and would see the error logs. But I don't think we should rely on how good a system admin of an Artemis instance is..